backend/src/main/java/com/redhat/cloud/notifications/events/KafkaMessageDeduplicator.java

```diff
@@ -0,0 +1,129 @@
+                        return messageId;
+                    } catch (IllegalArgumentException e) {
+                        invalidMessageIdCounter.increment();
+                        LOGGER.warnf("Application %s/%s sent an invalid Kafka header [%s=%s]. They must change their " +
+                                "integration and send a valid UUID.", bundleName, applicationName, MESSAGE_ID_HEADER, headerValue);
+                    }
+                }
+            }
+            if (headers.hasNext()) {
+                LOGGER.warnf("Application %s/%s sent multiple Kafka headers [%s]. They must change their " +
+                                "integration and send only one value.", bundleName, applicationName, MESSAGE_ID_HEADER);
+            }
+        }
+        if (!found) {
+            missingMessageIdCounter.increment();
+            LOGGER.tracef("Application %s/%s did not send any Kafka header [%s]",
+                    bundleName, applicationName, MESSAGE_ID_HEADER);
+        }
+        // This will be returned if the message ID is either invalid or not found.
+        return null;
+    }
+
+    /**
+     * Verifies that another Kafka consumer didn't already process the given message and then failed to commit its
+     * offset. Such failure can happen when a consumer is kicked out of its consumer group because it didn't poll new
+     * messages fast enough. We experienced that already on production.
+     */
+    public Uni<Boolean> isDuplicate(String bundleName, String applicationName, UUID messageId) {
+        if (messageId == null) {
+            /*
+             * For now, messages without an ID are always considered new. This is necessary to give the onboarded apps
+             * time to change their integration and start sending the new header. The message ID may become mandatory later.
+             */
+            return Uni.createFrom().item(FALSE);
+        } else {
+            String hql = "SELECT TRUE FROM KafkaMessage WHERE id.bundleName = :bundleName " +
+                    "AND id.applicationName = :applicationName AND id.messageId = :messageId";
+            return statelessSession.createQuery(hql, Boolean.class)
+                    .setParameter("bundleName", bundleName)
+                    .setParameter("applicationName", applicationName)
+                    .setParameter("messageId", messageId)
+                    .getSingleResultOrNull()
+                    .onItem().ifNull().continueWith(FALSE);
+        }
+    }
+
+    public Uni<Void> registerMessageId(String bundleName, String applicationName, UUID messageId) {
+        if (messageId == null) {
+            return Uni.createFrom().voidItem();
+        } else {
+            KafkaMessage kafkaMessage = new KafkaMessage(bundleName, applicationName, messageId);
+            kafkaMessage.prePersist(); // This method must be called manually while using a StatelessSession.
+            return statelessSession.insert(kafkaMessage);
```

Comment:
Are we cleaning those out?
I think we only need to keep them for a relatively short amount of time (1 day max)

Reply:
Yes, that's part of the plan (it's not committed yet though).

Reply:
Cleaner added.